source/simulator/introspection.cc

```diff
@@ -32,166 +32,157 @@ namespace aspect
 {
   namespace internal
   {
-
     /**
-     * Return pair with @p n_components and a filled ComponentIndices structure.
+     * Return a filled ComponentIndices structure.
      */
     template <int dim>
-    std::pair<unsigned int, typename Introspection<dim>::ComponentIndices>
-    setup_component_indices (const unsigned int n_compositional_fields)
+    typename Introspection<dim>::ComponentIndices
+    setup_component_indices (dealii::FEVariableSystem<dim> &ib)
     {
       typename Introspection<dim>::ComponentIndices ci;
-      unsigned int comp = 0;
-
       for (unsigned int i=0; i<dim; ++i)
-        ci.velocities[i] = comp++;
-      ci.pressure = comp++;
-      ci.temperature = comp++;
+        ci.velocities[i] = ib.variable("velocity").first_component_index+i;
+
+      ci.pressure = ib.variable("pressure").first_component_index;
+      ci.temperature = ib.variable("temperature").first_component_index;
+      unsigned int n_compositional_fields = ib.variable("compositions").n_components();
       for (unsigned int i=0; i<n_compositional_fields; ++i)
-        ci.compositional_fields.push_back(comp++);
-      return std::make_pair(comp, ci);
+        ci.compositional_fields.push_back(ib.variable("compositions").first_component_index+i);
```

Comment:
What happens if someone uses the signal to remove one of these default variables, e.g. the pressure? Then all of these `setup_...()` function will trigger an assertion. Is this what we want (we do not allow these variables to be removed) or should we handle this somehow?